### PR TITLE
chore(broker-core): extract event subscription logic

### LINKED ISSUE #1968: Streamline handlers of ExecutableCatchEventSupplier
After #1939, there is some duplication in the different element/lifecycle state handlers, primarily when it comes to elements that implement `ExecutableCatchEventSupplier`. We should streamline their behaviours such that there could be a `CatchEventSupplierElementActivatingHandler` as the parent of `ActivityElementActivatingHandler`, `EventBasedGatewayElementActivatingHandler`, etc.

### PR DESCRIPTION
**Pull request by npepinpe**
closes #1968

### CODE REVIEW

**Comment by npepinpe** [comment]
Most of the PR is unit tests for the single element-lifecycle tests.

One thing to note regarding the original issue, I decided against moving towards adding generations to the class hierarchy and try to move towards composition; there's a few (major) problems with what I did when I added the element state handler hierarchy, but with no time to fix them at once, I figured little by little (i.e. adding a `CatchEventSubscriber` instead of a `CatchEventSupplierElementActivatingHandler` in the hierarchy) woud be best.

**Review by saig0 — CHANGES_REQUESTED** [review]
Just small things. But I would like to discuss if we should write this kind of workflow processing tests.

File: broker-core/src/main/java/io/zeebe/broker/workflow/processor/handlers/activity/ActivityEventOccurredHandler.java [hunk 1]
```diff
@@ -61,7 +60,11 @@ protected boolean handleState(BpmnStepContext<T> context) {
       publishEvent(context, context.getRecord().getKey(), eventRecord, event);
     }
 
-    return super.handleState(context);
+    return true;
+  }
+
+  private boolean isActivityEventHandler(BpmnStepContext<T> context, EventTrigger event) {
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply broker-core/src/main/java/io/zeebe/broker/workflow/processor/handlers/activity/ActivityEventOccurredHandler.java]: My bad, it was moved to another class and there it's called.

File: broker-core/src/main/java/io/zeebe/broker/workflow/processor/handlers/activity/ActivityElementCompletingHandler.java [hunk 2]
```diff
@@ -19,33 +19,37 @@
...
   @Override
   protected boolean handleState(BpmnStepContext<T> context) {
-    if (super.handleState(context)) {
-      context.getCatchEventBehavior().unsubscribeFromEvents(context.getRecord().getKey(), context);
-      return true;
-    }
-
-    return false;
+    catchEventSubscriber.unsubscribeFromEvents(context);
+    return super.handleState(context);
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply broker-core/src/main/java/io/zeebe/broker/workflow/processor/handlers/activity/ActivityElementCompletingHandler.java]: Let's discuss that as well tomorrow in the office.

**Review by saig0 — APPROVED** [review]

**Comment by npepinpe** [comment]
bors retry

**Pull request merged**